### PR TITLE
Infer operation data type based on its params for Corr2D

### PR DESCRIPTION
**Pull request by ArtemSkrebkov**
# Motivation
* Extend Corr2D to infer operation data type based on its parameters
* Solves https://github.com/buddy-compiler/buddy-mlir/issues/21

# Opens
* Is it feasible to make output to be a returned value?
  * I gave a try but had to get back to a parameter for the output since I didn't manage to expose returned value via C interface

### CODE REVIEW

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp [hunk 1]
```diff
@@ -490,7 +508,7 @@ class DIPCorr2DOpLowering : public OpRewritePattern<dip::Corr2DOp> {
                 builder.create<scf::YieldOp>(loc);
               });
         });
-    // Remove the origin convolution operation.
+
```
Comment (ArtemSkrebkov) [inline lib/Conversion/LowerDIP/LowerDIPPass.cpp:511]: Not required.

File: lib/Dialect/DIP/DIPOps.cpp [hunk 2]
```diff
@@ -20,7 +20,10 @@
 
 #include "DIP/DIPOps.h"
 #include "DIP/DIPDialect.h"
+#include "mlir/IR/BuiltinAttributes.h"
 #include "mlir/IR/OpImplementation.h"
+#include "mlir/Support/LogicalResult.h"
 
 #define GET_OP_CLASSES
 #include "DIP/DIPOps.cpp.inc"
+
```
Comment (ArtemSkrebkov) [inline lib/Dialect/DIP/DIPOps.cpp:29]: Not required.

File: include/Utils/DIPUtils.h [hunk 3]
```diff
@@ -84,8 +98,15 @@ void calcAndStoreFMAwTailProcessing(OpBuilder &builder, Location loc,
...
-        Value resVec =
-            builder.create<FMAOp>(loc, inputVec, kernelVec, outputVec);
+        Value resVec = {};
+        auto elemTy = vecType.getElementType();
+        if (elemTy.isF32()) {
+          resVec = builder.create<vector::FMAOp>(loc, inputVec, kernelVec, outputVec);
+        } else if (elemTy.isInteger(32)) {
+          Value mulVec = builder.create<arith::MulIOp>(loc, inputVec, kernelVec);
+          resVec = builder.create<arith::AddIOp>(loc, mulVec, outputVec);
+        }
```
Comment (meshtag) [inline include/Utils/DIPUtils.h:108]: All of this (101 - 108) can be wrapped up nicely in a utility function for convenience

**Review by meshtag — CHANGES_REQUESTED** [review]
Thanks a lot for introducing llvm style tests for the `DIP` dialect!

Overall, I don't see any major issue with the implementation mechanism and I think you can continue with the rest of the implementation for covering remaining types.

File: tests/Dialect/DIP/correlation2D_invalid_type.mlir [hunk 4]
```diff
@@ -0,0 +1,61 @@
+//
+// x86
+//
+// RUN: not buddy-opt %s -lower-dip="DIP-strip-mining=64" -arith-expand --convert-vector-to-scf --lower-affine --convert-scf-to-cf --convert-vector-to-llvm \
```
[the remark this answers is not in the document]
Comment (ArtemSkrebkov) [reply tests/Dialect/DIP/correlation2D_invalid_type.mlir:4]: Since it is a negative test, `buddy-opt` will return error exit code and `FileCheck` will propagate what will make our test to fail.

So we need to add `not`, so `FileCheck` won't complain about error code. Looks like it is a common pattern for testing error sitatuion in `mlir`. One more example:
https://github.com/llvm/llvm-project/blob/14d79afeae63d78de9483f750fafaba13c7ae2dc/mlir/test/IR/diagnostic-handler.mlir

One more thing is that you need to check not only `stdout` as usual, but `stderr` as well (done by `2>&1`)

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp [hunk 5]
```diff
@@ -72,7 +73,17 @@ class DIPCorr2DOpLowering : public OpRewritePattern<dip::Corr2DOp> {
     auto boundaryOptionAttr = op.boundary_option();
     Value strideVal = rewriter.create<ConstantIndexOp>(loc, stride);
 
-    FloatType f32 = FloatType::getF32(ctx);
+    auto inElemTy = input.getType().cast<MemRefType>().getElementType();
+    auto kElemTy = kernel.getType().cast<MemRefType>().getElementType();
+    auto outElemTy = output.getType().cast<MemRefType>().getElementType();
+    auto constElemTy = constantValue.getType();
+    if (inElemTy != kElemTy || kElemTy != outElemTy || outElemTy != constElemTy) {
+        return op->emitOpError() <<  "input, kernel, output and constant must have the same element type";
```
[the remark this answers is not in the document]
Comment (ArtemSkrebkov) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp:81]: That makes sense. I will explore this.
However, likely, it would introduce some type conversion ops and bound checking, so it is better to make as a separate PR.

File: tests/Dialect/DIP/correlation2D_invalid_type.mlir (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArtemSkrebkov) [reply tests/Dialect/DIP/correlation2D_invalid_type.mlir:4]: Looks like `not` is not really required. I can remove it safely and the test passes also.

File: include/Utils/DIPUtils.h (hunk 3 above)
Comment (ArtemSkrebkov) [reply include/Utils/DIPUtils.h]: Added

**Comment by ArtemSkrebkov** [comment]
@meshtag 

Please take a look one more time. `f64` test might fail. I will fix it. But the other is ready for review.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArtemSkrebkov) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp]: So, since it requires adding a parameter (similar to opencv). I'd suggest making a separate issue to support it. I can create one. Is it ok?

File: include/Utils/DIPUtils.h [hunk 6]
```diff
@@ -24,6 +24,21 @@
 
 #include "Utils/Utils.h"
 
+Value insertFMAOp(OpBuilder &builder, Location loc, VectorType type,
+                  Value input, Value kernel, Value output) {
```
Comment (meshtag) [inline include/Utils/DIPUtils.h:28]: Please use `inputVec`, `'kernelVec`, `outputVec` here for clarity.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp [hunk 7]
```diff
@@ -44,6 +48,23 @@ using namespace mlir::arith;
 
 namespace {
 
+Value insertZeroConstantOp(MLIRContext *ctx, OpBuilder &builder, Location loc, Type elemTy) {
```
Comment (meshtag) [inline lib/Conversion/LowerDIP/LowerDIPPass.cpp:51]: I think this function can be suitably placed inside `Utils.h`

File: lib/Dialect/DIP/DIPOps.cpp [hunk 8]
```diff
@@ -24,3 +24,4 @@
...
```
Comment (meshtag) [inline lib/Dialect/DIP/DIPOps.cpp:27]: Unnecessary whitespace?

**Review by meshtag — CHANGES_REQUESTED** [review]
I'll build this branch and have a more detailed look once the issue with `f64` test is fixed.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (meshtag) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp]: Okay

File: include/Utils/DIPUtils.h (hunk 6 above)
Comment (ArtemSkrebkov) [reply include/Utils/DIPUtils.h]: Alright.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp (hunk 7 above)
Comment (ArtemSkrebkov) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp]: Moved.

File: lib/Dialect/DIP/DIPOps.cpp (hunk 8 above)
Comment (ArtemSkrebkov) [reply lib/Dialect/DIP/DIPOps.cpp]: Removed.

**Comment by ArtemSkrebkov** [comment]
`f64` can work as well, so ready for one more round of review.

File: tests/Dialect/DIP/correlation2D_f32.mlir [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+func.func @main() -> i32 {
+  %input = memref.get_global @global_input : memref<3x3xf32>
+  %identity = memref.get_global @global_identity : memref<3x3xf32>
+  %output = memref.get_global @global_output : memref<3x3xf32>
+
+  %x = arith.constant 1 : index
+  %y = arith.constant 1 : index
+  %c = arith.constant 0. : f32 
+  dip.corr_2d <CONSTANT_PADDING> %input, %identity, %output, %x, %x, %c : memref<3x3xf32>, memref<3x3xf32>, memref<3x3xf32>, index, index, f32
```
Comment (meshtag) [inline tests/Dialect/DIP/correlation2D_f32.mlir:32]: I think you wanted to pass `%y` as y co-ordinate for kernel anchor point, right?
Also, please consider naming `%x`, `%y` to `%kernelAnchorX`, `%kernelAnchorY` respectively.

Please consider the above point for all other tests as well.

File: tests/Dialect/DIP/correlation2D_invalid_type.mlir [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+  dip.corr_2d <CONSTANT_PADDING> %input_f32, %identity_i32, %output_f32, %x, %x, %c_f32 : memref<3x3xf32>, memref<3x3xi32>, memref<3x3xf32>, index, index, f32
+  // CHECK: 'dip.corr_2d' op input, kernel, output and constant must have the same element type
+
+  dip.corr_2d <CONSTANT_PADDING> %input_f32, %identity_f32, %output_i32, %x, %x, %c_f32 : memref<3x3xf32>, memref<3x3xf32>, memref<3x3xi32>, index, index, f32
+  // CHECK: 'dip.corr_2d' op input, kernel, output and constant must have the same element type
+
+  dip.corr_2d <CONSTANT_PADDING> %input_f32, %identity_f32, %output_f32, %x, %x, %c_i32 : memref<3x3xf32>, memref<3x3xf32>, memref<3x3xf32>, index, index, i32
+  // CHECK: 'dip.corr_2d' op input, kernel, output and constant must have the same element type
+
```
Comment (meshtag) [inline tests/Dialect/DIP/correlation2D_invalid_type.mlir:58]: Please add a test for [this](https://github.com/ArtemSkrebkov/buddy-mlir/blob/7d7a8db947e703fd3ae2229da3b5aa19de64f763/lib/Conversion/LowerDIP/LowerDIPPass.cpp#L109) condition as well.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp [hunk 11]
```diff
@@ -44,6 +48,23 @@ using namespace mlir::arith;
 
 namespace {
 
+Value insertZeroConstantOp(MLIRContext *ctx, OpBuilder &builder, Location loc,
```
Comment (meshtag) [inline lib/Conversion/LowerDIP/LowerDIPPass.cpp:51]: Please move this function inside `Utils.h`. Also, please add a comment line explaining the motive of this function, similar to other functions in `Utils.h`.

File: include/Utils/DIPUtils.h [hunk 12]
```diff
@@ -24,6 +24,21 @@
 
 #include "Utils/Utils.h"
 
+Value insertFMAOp(OpBuilder &builder, Location loc, VectorType type,
```
Comment (meshtag) [inline include/Utils/DIPUtils.h:27]: Please add a comment line explaining the motive of this function, similar to other functions in `DIPUtils.h`.

**Review by meshtag — CHANGES_REQUESTED** [review]

File: include/Utils/DIPUtils.h (hunk 12 above)
Comment (ArtemSkrebkov) [reply include/Utils/DIPUtils.h:50]: Added.

File: tests/Dialect/DIP/correlation2D_f32.mlir (hunk 9 above)
Comment (ArtemSkrebkov) [reply tests/Dialect/DIP/correlation2D_f32.mlir]: Done.

File: tests/Dialect/DIP/correlation2D_invalid_type.mlir (hunk 10 above)
Comment (ArtemSkrebkov) [reply tests/Dialect/DIP/correlation2D_invalid_type.mlir:75]: Updated `correlation2D_invalid_type.mlir` to check the condition as well.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp (hunk 11 above)
Comment (ArtemSkrebkov) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp]: Done.

**Review by meshtag — APPROVED** [review]
LGTM! (/cc @zhanghb97 )
Thanks @ArtemSkrebkov !

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (meshtag) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp]: @ArtemSkrebkov , can you please create a feature request so that we can keep track of this potential enhancement.

[the remark this answers is not in the document]
Comment (ArtemSkrebkov) [reply lib/Conversion/LowerDIP/LowerDIPPass.cpp]: Sure, done
https://github.com/buddy-compiler/buddy-mlir/issues/66

Seems, I am not able to set labels, so feel free to add some.

**Comment by zhanghb97** [comment]
Thanks @ArtemSkrebkov @meshtag !
